src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -194,11 +194,13 @@ public void OnCompleted(Action continuation)
             }
             else if (awaitableState == _awaitableIsCompleted)
             {
-                ThreadPool.QueueUserWorkItem((o) => ((Action)o)(), continuation);
+                // Read already completed, continue on calling thread
+                continuation();
```

Comment:
No; just came back to this code investigating why I hit the 3rd state neither IsNotCompleted or IsCompleted and the stack trace showed top of stack. However I only hit this 3rd state 3 times in 30 mins under load; and as you point out it being completed in the interval is probably a rare state.